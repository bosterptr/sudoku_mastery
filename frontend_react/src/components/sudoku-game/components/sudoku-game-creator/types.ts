// eslint-disable-next-line no-shadow
export const enum StepType {
  SetNumber = 0,
  SetNote = 1,
}
export const SudokuStageOfTheGame = {
  INITIAL: 0,
  WON: 1,
  LOST: 2,
} as const;
export type ISudokuStageOfTheGame =
  (typeof SudokuStageOfTheGame)[keyof typeof SudokuStageOfTheGame];
export type ISudokuValues = string;
// 0 index based
export type INotes = number[];
/** e.g. '3.8.15...4.97....5561..8...9.4...53..5...1.49......8...3...7..67..39.2......8.7..' */
export type IMission = string;
/** XXY = XX-cell index, Y-value. e.g., '205' represents cell index 20 and value 5. */
export type IScoreHistory = string[];
export interface IStep {
  index: number;
  previous: string;
  type: StepType;
  value: string;
}
export type ICellNotes = number[];
