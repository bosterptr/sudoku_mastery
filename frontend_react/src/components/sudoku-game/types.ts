// eslint-disable-next-line no-shadow
export const enum StepType {
  SetNumber = 0,
  SetNote = 1,
}
export const SudokuStageOfTheGame = {
  INITIAL: 0,
  WON: 1,
  LOST: 2,
} as const;
export type ISudokuStageOfTheGame =
  (typeof SudokuStageOfTheGame)[keyof typeof SudokuStageOfTheGame];
export type ISudokuValues = string;
// 0 index based
export type INotes = number[];
// eslint-disable-next-line no-shadow
export const SudokuDifficulty = {
  Easy: 'easy',
  Medium: 'normal',
  Hard: 'hard',
  Expert: 'extreme',
} as const;
export interface ISudoku {
  id: string;
  body: string;
  difficulty: (typeof SudokuDifficulty)[keyof typeof SudokuDifficulty];
  created_at: string;
  updated_at: string;
  user_id: string;
}
/** e.g. '3.8.15...4.97....5561..8...9.4...53..5...1.49......8...3...7..67..39.2......8.7..' */
export type IMission = string;
/** XXY = XX-cell index, Y-value. e.g., '205' represents cell index 20 and value 5. */
export type IScoreHistory = string[];
export interface IStep {
  index: number;
  previous: string;
  type: StepType;
  value: string;
}
export type ICellNotes = number[];
