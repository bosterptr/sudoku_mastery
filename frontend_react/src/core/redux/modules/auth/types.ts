export interface SessionUser {
  id: string;
  email: string;
  displayName: string;
}

export interface APIState {
  done: boolean;
  hasErrored: boolean;
  isLoading: boolean;
}
// eslint-disable-next-line no-shadow
export const AuthAPIStateKeys = {
  loginWithEmail: 0,
  changePassword: 1,
  changePasswordWithToken: 2,
  activateAccount: 3,
  forgotPassword: 4,
} as const;
export type IAuthAPIStateKeys = (typeof AuthAPIStateKeys)[keyof typeof AuthAPIStateKeys];
export interface AuthAPIState {
  [AuthAPIStateKeys.loginWithEmail]?: APIState;
  [AuthAPIStateKeys.changePassword]?: APIState;
  [AuthAPIStateKeys.changePasswordWithToken]?: APIState;
  [AuthAPIStateKeys.activateAccount]?: APIState;
  [AuthAPIStateKeys.forgotPassword]?: APIState;
}
export type IAuthState = AuthAPIState;
