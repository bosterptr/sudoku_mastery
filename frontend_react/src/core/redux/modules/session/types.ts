export interface SessionUser {
  id: string;
  email: string;
  displayName: string;
}

export interface ISessionState {
  firstPageLoaded: boolean;
  user: SessionUser | null;
}
