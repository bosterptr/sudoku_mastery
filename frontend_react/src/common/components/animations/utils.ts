import { CSSProperties } from 'react';

export const reflow = (node: Element) => node.scrollTop;

interface ComponentProps {
  easing: string | { enter?: string; exit?: string } | undefined;
  style: CSSProperties | undefined;
  timeout: number | { enter?: number; exit?: number };
}

interface Options {
  mode: 'enter' | 'exit';
}

interface TransitionProps {
  duration: string | number;
  easing: string | undefined;
  delay: string | undefined;
}

export function getTransitionProps(props: ComponentProps, options: Options): TransitionProps {
  const { timeout, easing, style = {} } = props;

  return {
    duration:
      style.transitionDuration ??
      (typeof timeout === 'number' ? timeout : timeout[options.mode] || 0),
    easing:
      style.transitionTimingFunction ??
      (typeof easing === 'object' ? easing[options.mode] : easing),
    delay: style.transitionDelay,
  };
}

export function getAutoHeightDuration(height: number) {
  if (!height) return 0;
  const constant = height / 36;
  // https://www.wolframalpha.com/input/?i=(4+%2B+15+*+(x+%2F+36+)+**+0.25+%2B+(x+%2F+36)+%2F+5)+*+10
  return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
}
