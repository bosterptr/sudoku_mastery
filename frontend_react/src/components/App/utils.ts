const GL_BUFFER_STATUS_CHECK_INTERVAL = 2;

export type IOpenglSetup = [
  OffscreenCanvas,
  WebGLProgram,
  WebGLProgram,
  WebGLProgram,
  ITexFbPair,
  ITexFbPair,
  number,
];

export interface ITexFbPair {
  tex: WebGLTexture;
  fb: WebGLFramebuffer;
}
export const getGrayScaleFromUint8Array = (
  dataArray: Uint8Array,
  width: number,
  height: number,
  cachedGrayScaleUintArray: Uint8Array,
  cachedGrayScaleSize: number,
): Uint8Array => {
  if (cachedGrayScaleSize !== width * height) {
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray = new Uint8Array(width * height);
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleSize = width * height;
  }
  const { length } = dataArray;
  for (let i = 0, luminanceIndex = 0; i < length; i += 4, luminanceIndex += 1) {
    // eslint-disable-next-line no-bitwise
    const r = dataArray[i] >> 2;
    // eslint-disable-next-line no-bitwise
    const g = dataArray[i + 1] >> 1;
    // eslint-disable-next-line no-bitwise
    const b = dataArray[i + 2] >> 3;
    // const luminance = 0.299 * r + 0.587 * g + 0.114 * b;
    // luminanceData[luminanceIndex] =
    //   0.299 * dataArray[i] +
    //   0.587 * dataArray[i + 1] +
    //   0.114 * dataArray[i + 2];
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray[luminanceIndex] = r + b + g;
  }
  return cachedGrayScaleUintArray;
};

export const clientWaitAsync = (
  gl: WebGL2RenderingContext,
  sync: WebGLSync,
  flags = 0,
  intervalMs = 10,
) => {
  return new Promise<void>((resolve, reject) => {
    const check = () => {
      const res = gl.clientWaitSync(sync, flags, 0);
      if (res === gl.WAIT_FAILED) {
        reject();
        return;
      }
      if (res === gl.TIMEOUT_EXPIRED) {
        setTimeout(check, intervalMs);
        return;
      }
      resolve();
    };
    check();
  });
};

export const readPixelsAsync = (
  gl: WebGL2RenderingContext,
  width: number,
  height: number,
  buffer: Uint8Array,
): Promise<Uint8Array> => {
  const bufpak = gl.createBuffer();
  gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
  gl.bufferData(gl.PIXEL_PACK_BUFFER, buffer.byteLength, gl.STREAM_READ);
  gl.readPixels(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, 0);
  const sync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
  if (!sync) throw new Error('WebGLSync is empty');
  gl.flush();
  return clientWaitAsync(gl, sync, 0, GL_BUFFER_STATUS_CHECK_INTERVAL).then(() => {
    gl.deleteSync(sync);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
    gl.getBufferSubData(gl.PIXEL_PACK_BUFFER, 0, buffer);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, null);
    gl.deleteBuffer(bufpak);
    return buffer;
  });
};

export function setAttributes(gl: WebGL2RenderingContext, program: WebGLProgram, name: string) {
  const location = gl.getAttribLocation(program, name);
  gl.enableVertexAttribArray(location);
  gl.vertexAttribPointer(location, 2, gl.FLOAT, false, 0, 0);
}

export function isMobile() {
  const userAgent = navigator.userAgent.toLowerCase();
  return /iphone|ipod|android|ie|blackberry|fennec/.test(userAgent);
}
export const getCameraResolution = () => {
  if (isMobile()) {
    return {
      width: { ideal: document.body.clientWidth > 500 ? 500 : document.body.clientWidth },
      height: { ideal: document.body.clientHeight > 500 ? 500 : document.body.clientHeight },
    };
  }
  return {
    width: { ideal: 1000 },
    height: { ideal: 1000 },
  };
};
