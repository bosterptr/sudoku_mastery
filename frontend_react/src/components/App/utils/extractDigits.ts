import { LayersModel, loadLayersModel, reshape, tensor2d } from '@tensorflow/tfjs';

const indexOfMaxValue = <T>(
  a: T[],
  func: (param: T) => number = ((i) => i) as (param: T) => number,
) => a.reduce((iMax, x, i, arr) => (func(x) > func(arr[iMax]) ? i : iMax), 0);

let model: LayersModel;
const CELLS_IN_ROW = 9;
const CELL_WIDTH = 20;
const TOTAL_CELLS = CELLS_IN_ROW * CELLS_IN_ROW;
const IMAGE_WIDTH = CELLS_IN_ROW * CELL_WIDTH;
const CELL_PIXEL_COUNT = CELL_WIDTH * CELL_WIDTH;

function extractCube(src: Uint8Array, x: number, y: number): Uint8Array {
  if (x < 0 || y < 0 || x + CELL_WIDTH > IMAGE_WIDTH || y + CELL_WIDTH > IMAGE_WIDTH) {
    throw new Error('Starting point out of bounds');
  }
  const cube = new Uint8Array(CELL_WIDTH * CELL_WIDTH);
  for (let row = 0; row < CELL_WIDTH; row += 1) {
    for (let col = 0; col < CELL_WIDTH; col += 1) {
      const srcIndex = (y + row) * IMAGE_WIDTH + (x + col);
      const cubeIndex = row * CELL_WIDTH + col;
      cube[cubeIndex] = src[srcIndex];
    }
  }

  return cube;
}

const init = async () => {
  model = await loadLayersModel('./models/model6.json');
};
const extractDigits = (cells: Uint8Array) => {
  const testDataArray = new Float32Array(IMAGE_WIDTH * IMAGE_WIDTH);
  for (let i = 0; i < TOTAL_CELLS; i += 1) {
    const x = (i % CELLS_IN_ROW) * CELL_WIDTH;
    const y = Math.floor(i / CELLS_IN_ROW) * CELL_WIDTH;
    const buffer = extractCube(cells, x, y);
    testDataArray.set(buffer, i * CELL_PIXEL_COUNT);
  }
  const testTensor = tensor2d(testDataArray, [TOTAL_CELLS, CELL_PIXEL_COUNT]);
  const reshaped = reshape(testTensor, [TOTAL_CELLS, CELL_WIDTH, CELL_WIDTH, 1]);
  const tensor = model.predict(reshaped);
  if (Array.isArray(tensor)) throw new Error();
  const prediction = tensor.dataSync();
  const predictionData: number[] = [];
  for (const num of prediction) {
    predictionData.push(num);
  }
  let result = '';
  for (let i = 0; i < TOTAL_CELLS; i += 1) {
    const cellPrediction = predictionData.slice(i * 10, i * 10 + 10);
    const digit = indexOfMaxValue(cellPrediction);
    result += digit === 0 ? '.' : digit;
  }
  return result;
};

export { extractDigits, init };
