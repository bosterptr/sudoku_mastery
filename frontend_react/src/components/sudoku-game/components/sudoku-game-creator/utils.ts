import { ALL_SUDOKU_CELLS, CANVAS_SIZE, SUDOKU_CELLS_IN_ROW } from 'app/core/constants/sudoku';
import { ICellNotes, IMission } from './types';

export const generateNotes = (): ICellNotes[] => {
  const newArray: ICellNotes[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    newArray.push([]);
  }
  return newArray;
};
export const getBlockSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 2 * pixelRatio) / 3;
export const getAdjustedPixelRatio = () => {
  const { devicePixelRatio } = window;
  return devicePixelRatio > 2 ? 2 * devicePixelRatio : 4;
};
export const getCellSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 4 * pixelRatio) / 9;
export const getContextSize = (pixelRatio: number) => CANVAS_SIZE * pixelRatio;
export const getOuterSize = (pixelRatio: number) => 2 * pixelRatio;
export const getThinSize = (pixelRatio: number) => 0.75 * pixelRatio;
export const getX = (index: number) => Math.floor(index / 9);
export const getY = (index: number) => index % 9;
export const getCellXPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getY(index);
export const getCellYPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getX(index);
export const getCubeId = (index: number) =>
  `${Math.floor(getX(index) / 3)}${Math.floor(getY(index) / 3)}`;
export const EMPTY_CELL = '.';
export const isSudokuCellEmpty = (char: string) => char === EMPTY_CELL;
export const generateArrWithTrueAtIndexes = (indexes: number[]): boolean[] => {
  const newArray: boolean[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (indexes.includes(cellIndex)) newArray[cellIndex] = true;
    else newArray[cellIndex] = false;
  }
  return newArray;
};
export const getScoreFromMove = (values: IMission, solution: string, selectedCell: number) => {
  let fullRows = 0;
  let fullColumns = 0;
  let fullCubes = 0;
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (values[cellIndex] === solution[cellIndex]) {
      if (getY(cellIndex) === getY(selectedCell)) fullRows += 1;
      if (getX(cellIndex) === getX(selectedCell)) fullColumns += 1;
      if (getCubeId(cellIndex) === getCubeId(selectedCell)) fullCubes += 1;
    }
  }
  let score = 1;
  if (fullRows === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullColumns === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullCubes === 9) score += 5;
  return score;
};
