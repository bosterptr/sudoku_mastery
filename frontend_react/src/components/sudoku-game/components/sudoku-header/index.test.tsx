/* eslint-disable react/jsx-props-no-spreading */
import { render } from 'app/tests/utils';
import { IPublicProps, SudokuHeader } from '.';
import { SudokuDifficulty } from '../../types';

const requiredProps: IPublicProps = {
  difficulty: SudokuDifficulty.Medium,
  mistakes: 2,
  onChangeGameType: jest.fn(),
  score: 2137,
  seconds: 0,
};

const allProps: Required<IPublicProps> = {
  ...requiredProps,
  className: 'x',
};

describe('SudokuHeader', () => {
  it('renders component with required props', () => {
    const { asFragment } = render(<SudokuHeader {...requiredProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders component with all props', () => {
    const { asFragment } = render(<SudokuHeader {...allProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
