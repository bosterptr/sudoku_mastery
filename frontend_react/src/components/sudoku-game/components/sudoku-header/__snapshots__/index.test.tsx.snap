// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SudokuHeader renders component with all props 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.c2 {
  font-size: 16px;
  font-weight: 500;
  color: #24292ebf;
}

.c3 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  padding: 6px 8px;
  font-size: 14px;
  cursor: pointer;
  color: #1648ff;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.c3:focus {
  outline: none;
  text-decoration: none;
}

.c3:hover {
  color: #fff;
  background-color: #1236bb14;
  text-decoration: none;
}

.c3:active {
  background-color: #1648ff;
  color: #fff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c4 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  padding: 6px 8px;
  font-size: 14px;
  cursor: pointer;
  color: #1648ff;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  background-color: #ffffff;
}

.c4:focus {
  outline: none;
  text-decoration: none;
}

.c4:hover {
  color: #fff;
  background-color: #1236bb14;
  text-decoration: none;
}

.c4:active {
  background-color: #1648ff;
  color: #fff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0 {
  align-items: center;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  margin-bottom: 5px;
  position: relative;
}

.c6 {
  font-weight: 400;
  font-size: 16px;
  line-height: 1.43;
  color: #3d4150;
}

.c5 {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 38%;
}

.c7 {
  color: #000000;
  font-size: 13px;
  font-weight: 600;
  margin-left: 4px;
}

<div
    class="c0 x"
  >
    <span
      class="c1"
    >
      <div
        class="c2"
      >
        Difficulty:
      </div>
      <div
        class="c3"
      >
        Easy
      </div>
      <div
        class="c4"
      >
        Medium
      </div>
      <div
        class="c3"
      >
        Hard
      </div>
      <div
        class="c3"
      >
        Expert
      </div>
    </span>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        Mistakes:
        <span
          class="c7"
        >
          2/3
        </span>
      </div>
      <div
        class="c6"
      >
        Score:
        <span
          class="c7"
        >
          2137
        </span>
      </div>
      <div
        class="c6"
      >
        00:00
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SudokuHeader renders component with required props 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.c2 {
  font-size: 16px;
  font-weight: 500;
  color: #24292ebf;
}

.c3 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  padding: 6px 8px;
  font-size: 14px;
  cursor: pointer;
  color: #1648ff;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.c3:focus {
  outline: none;
  text-decoration: none;
}

.c3:hover {
  color: #fff;
  background-color: #1236bb14;
  text-decoration: none;
}

.c3:active {
  background-color: #1648ff;
  color: #fff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c4 {
  align-items: center;
  border-radius: 3px;
  border: none;
  box-shadow: none;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-weight: 500;
  justify-content: center;
  line-height: 1.75;
  text-decoration: none;
  transition: color 0.2s ease,background-color 0.2s ease,box-shadow 0.2s ease,border-color 0.2s ease,filter 0.2s ease;
  user-select: none;
  padding: 6px 8px;
  font-size: 14px;
  cursor: pointer;
  color: #1648ff;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  background-color: #ffffff;
}

.c4:focus {
  outline: none;
  text-decoration: none;
}

.c4:hover {
  color: #fff;
  background-color: #1236bb14;
  text-decoration: none;
}

.c4:active {
  background-color: #1648ff;
  color: #fff;
  box-shadow: 0px 4px 8px #181a1b33;
}

.c0 {
  align-items: center;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  margin-bottom: 5px;
  position: relative;
}

.c6 {
  font-weight: 400;
  font-size: 16px;
  line-height: 1.43;
  color: #3d4150;
}

.c5 {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 38%;
}

.c7 {
  color: #000000;
  font-size: 13px;
  font-weight: 600;
  margin-left: 4px;
}

<div
    class="c0"
  >
    <span
      class="c1"
    >
      <div
        class="c2"
      >
        Difficulty:
      </div>
      <div
        class="c3"
      >
        Easy
      </div>
      <div
        class="c4"
      >
        Medium
      </div>
      <div
        class="c3"
      >
        Hard
      </div>
      <div
        class="c3"
      >
        Expert
      </div>
    </span>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        Mistakes:
        <span
          class="c7"
        >
          2/3
        </span>
      </div>
      <div
        class="c6"
      >
        Score:
        <span
          class="c7"
        >
          2137
        </span>
      </div>
      <div
        class="c6"
      >
        00:00
      </div>
    </div>
  </div>
</DocumentFragment>
`;
